batch/batch/driver/scheduler.py

```diff
@@ -111,33 +120,143 @@ def allocate_cores(user, mark):
         while True:
             self.scheduler_state_changed.set()
-            self.cancel_state_changed.set()
+            self.cancel_ready_state_changed.set()
+            self.cancel_running_state_changed.set()
             await asyncio.sleep(60)
 
-    async def cancel_loop_body(self):
-        user_records = self.db.select_and_fetchall(
+    async def cancel_cancelled_ready_jobs_loop_body(self):
+        records = self.db.select_and_fetchall(
             '''
-SELECT user
-FROM user_resources
-GROUP BY user
-HAVING COALESCE(SUM(running_cores_mcpu), 0) > 0;
+SELECT user, n_cancelled_ready_jobs
+FROM (SELECT user,
+    CAST(COALESCE(SUM(n_cancelled_ready_jobs), 0) AS SIGNED) AS n_cancelled_ready_jobs
+  FROM user_resources
+  GROUP BY user) AS t
+WHERE n_cancelled_ready_jobs > 0;
 ''')
+        user_n_cancelled_ready_jobs = {
+            record['user']: record['n_cancelled_ready_jobs'] async for record in records
+        }
+
+        total = sum(user_n_cancelled_ready_jobs.values())
+        user_share = {
+            user: max(int(1000 * user_n_jobs / total + 0.5), 20)
+            for user, user_n_jobs in user_n_cancelled_ready_jobs.items()
+        }
+
+        async def user_cancelled_ready_jobs(user, remaining):
+            async for batch in self.db.select_and_fetchall(
+                    '''
+SELECT id, cancelled
+FROM batches
+WHERE user = %s AND `state` = 'running';
+''',
+                    (user,)):
+                if batch['cancelled']:
+                    async for record in self.db.select_and_fetchall(
+                            '''
+SELECT jobs.job_id
+FROM jobs
+WHERE batch_id = %s AND always_run = 0 AND state = 'Ready'
+LIMIT %s;
+''',
+                            (batch['id'], remaining.value)):
+                        record['batch_id'] = batch['id']
+                        yield record
+                else:
+                    async for record in self.db.select_and_fetchall(
+                            '''
+SELECT jobs.job_id
+FROM jobs
+WHERE batch_id = %s AND always_run = 0 AND state = 'Ready' AND cancelled = 1
+LIMIT %s;
+''',
+                            (batch['id'], remaining.value)):
+                        record['batch_id'] = batch['id']
+                        yield record
 
         should_wait = True
         async_work = []
-        async for user_record in user_records:
-            records = self.db.execute_and_fetchall(
-                '''
-SELECT attempts.job_id, attempts.batch_id, attempts.attempt_id, instance_name
-FROM attempts
-STRAIGHT_JOIN batches ON batches.id = attempts.batch_id
-STRAIGHT_JOIN jobs ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
-WHERE jobs.state = 'Running' AND (NOT jobs.always_run) AND batches.closed AND batches.cancelled AND batches.user = %s
-LIMIT 50;
+        for user, share in user_share.items():
+            remaining = Box(share)
+            async for record in user_cancelled_ready_jobs(user, remaining):
```

Comment:
Because I want to refer to remaining on each job query (which are called asynchronously) and remaining is changing continuously.  I can't pass it in as an argument as-is because that will capture the initial value.

I actually tried to write it so the definition of `user_cancelled_ready_jobs` was lexically below the definition of remaining and I could just reference remaining but I was getting a pylint error: Cell variable defined in loop (cell-var-from-loop).  I think the code was still correct (I don't totally understand that warning), fixed the warning and actually seemed clearer overall.

Reply:
This warning is intended to prevent this mistake:
```
incrementFuns = []
for i in range(0, 10):
    def addN(x):
        return i + x
    incrementFuns.append(addN)
```
wherein the developer thinks they've captured `i` by value.